File: docs/features/active/2026-08-17-orchestrator-remediation-loop-control-484/evidence/baseline/typescript-benchmark-build/lib/validate/orchestrator-state-codex-topology.js
"use strict";
/** Deterministic Codex topology receipt validation and delegation gates. */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CODEX_TOPOLOGY_RECEIPTS_KEY = void 0;
exports.validateCodexTopologyReceipt = validateCodexTopologyReceipt;
exports.validateCodexTopologyReceipts = validateCodexTopologyReceipts;
exports.validateCodexTopologyGate = validateCodexTopologyGate;
exports.validateCodexTopologyState = validateCodexTopologyState;
const codex_topology_resolver_1 = require("./codex-topology-resolver");
const orchestrator_state_codex_model_routing_1 = require("./orchestrator-state-codex-model-routing");
exports.CODEX_TOPOLOGY_RECEIPTS_KEY = "codex_topology_receipts";
const REQUIRED_KEYS = [
    "phase",
    "execution_context",
    "languages",
    "production_file_count",
    "test_file_count",
    "cross_cutting",
    "root_persona",
    "route",
    "topology",
    "logical_agent",
    "routing_reason",
    "max_production_files",
    "max_test_files",
];
const RESOLVED_KEYS = REQUIRED_KEYS.filter((key) => key !== "phase");
const FORCED_ROOT_PERSONAS = new Set([
    "epic-orchestrator",
    "epic-planner",
]);
function isObject(value) {
    return typeof value === "object" && value !== null && !Array.isArray(value);
}
function pythonRepr(value) {
    if (value === undefined || value === null) {
        return "None";
    }
    if (typeof value === "string") {
        const escaped = value
            .replaceAll("\\", "\\\\")
            .replaceAll("'", "\\'")
            .replaceAll("\n", "\\n")
            .replaceAll("\r", "\\r")
            .replaceAll("\t", "\\t");
        return `'${escaped}'`;
    }
    if (typeof value === "boolean") {
        return value ? "True" : "False";
    }
    if (Array.isArray(value)) {
        return `[${value.map(pythonRepr).join(", ")}]`;
    }
    if (isObject(value)) {
        const entries = Object.entries(value).map(([key, item]) => `${pythonRepr(key)}: ${pythonRepr(item)}`);
        return `{${entries.join(", ")}}`;
    }
    return String(value);
}
function valuesEqual(actual, expected) {
    if (Array.isArray(actual) && Array.isArray(expected)) {
        return (actual.length === expected.length &&
            actual.every((item, index) => item === expected[index]));
    }
    return actual === expected;
}
function validateReceiptInputs(receipt, prefix) {
    const errors = [];
    const languages = receipt["languages"];
    if (!Array.isArray(languages) ||
        languages.some((language) => typeof language !== "string" || language.trim().length === 0)) {
        errors.push(`${prefix}.languages must be a list of non-empty strings.`);
    }
    for (const key of ["production_file_count", "test_file_count"]) {
        const value = receipt[key];
        if (typeof value !== "number" || !Number.isInteger(value)) {
            errors.push(`${prefix}.${key} must be an integer.`);
        }
    }
    if (typeof receipt["cross_cutting"] !== "boolean") {
        errors.push(`${prefix}.cross_cutting must be a boolean.`);
    }
    if (typeof receipt["execution_context"] !== "string") {
        errors.push(`${prefix}.execution_context must be a string.`);
    }
    const rootPersona = receipt["root_persona"];
    if (rootPersona !== null && !FORCED_ROOT_PERSONAS.has(rootPersona)) {
        errors.push(`${prefix}.root_persona must be null or one of ` +
            "('epic-orchestrator', 'epic-planner').");
    }
    return errors;
}
function validateReceipt(value, prefix) {
    if (!isObject(value)) {
        return [`${prefix} must be an object.`];
    }
    const missing = REQUIRED_KEYS.filter((key) => !(key in value));
    if (missing.length > 0) {
        return [`${prefix} missing required keys: ${missing.join(", ")}.`];
    }
    const errors = [];
    const phase = value["phase"];
    if (typeof phase !== "string" || phase.trim().length === 0) {
        errors.push(`${prefix}.phase must be a non-empty string.`);
    }
    const inputErrors = validateReceiptInputs(value, prefix);
    errors.push(...inputErrors);
    if (inputErrors.length > 0) {
        return errors;
    }
    let expected;
    try {
        expected = (0, codex_topology_resolver_1.resolveCodexTopology)(value["languages"], value["production_file_count"], value["test_file_count"], value["execution_context"], {
            crossCutting: value["cross_cutting"],
            rootPersona: value["root_persona"],
        });
    }
    catch (error) {
        const message = error instanceof Error ? error.message : String(error);
        errors.push(`${prefix} has invalid routing inputs: ${message}`);
        return errors;
    }
    for (const key of RESOLVED_KEYS) {
        if (!valuesEqual(value[key], expected[key])) {
            errors.push(`${prefix}.${key} must be ${pythonRepr(expected[key])}, ` +
                `found ${pythonRepr(value[key])}.`);
        }
    }
    return errors;
}
/** Validate one receipt with a caller-selected diagnostic prefix. */
function validateCodexTopologyReceipt(value, prefix = `Checkpoint ${exports.CODEX_TOPOLOGY_RECEIPTS_KEY}[0]`) {
    return validateReceipt(value, prefix);
}
/** Validate every present topology receipt against the canonical resolver. */
function validateCodexTopologyReceipts(value) {
    if (!Array.isArray(value)) {
        return [
            `Checkpoint ${exports.CODEX_TOPOLOGY_RECEIPTS_KEY} must be a list when present.`,
        ];
    }
    return value.flatMap((item, index) => validateReceipt(item, `Checkpoint ${exports.CODEX_TOPOLOGY_RECEIPTS_KEY}[${index}]`));
}
function delegatedAgentNames(state) {
    const result = new Set();
    let receipts = state["delegation_receipts"];
    if (isObject(receipts)) {
        receipts = receipts["agents"];
    }
    if (!Array.isArray(receipts)) {
        return result;
    }
    for (const item of receipts) {
        if (!isObject(item)) {
            continue;
        }
        const agentName = item["agent_name"];
        if (typeof agentName === "string" && agentName.trim().length > 0) {
            result.add(agentName);
        }
    }
    return result;
}
function validReceipts(value) {
    return value.filter((item) => isObject(item) && validateCodexTopologyReceipts([item]).length === 0);
}
function modelDeployments(state, logicalAgent) {
    const result = new Set();
    const receipts = state[orchestrator_state_codex_model_routing_1.CODEX_MODEL_ROUTING_RECEIPTS_KEY];
    if (!Array.isArray(receipts)) {
        return result;
    }
    for (const item of receipts) {
        if (!isObject(item) ||
            (0, orchestrator_state_codex_model_routing_1.validateCodexModelRoutingReceipt)(item).length > 0 ||
            item["logical_agent"] !== logicalAgent) {
            continue;
        }
        const deployment = item["deployment_agent"];
        if (typeof deployment === "string") {
            result.add(deployment);
        }
    }
    return result;
}
/** Require exact topology evidence for root personas and initial delegates. */
function validateCodexTopologyGate(state, options = {}) {
    const delegated = delegatedAgentNames(state);
    const requiredRootPersona = options.requiredRootPersona;
    if (delegated.size === 0 && requiredRootPersona === undefined) {
        return [];
    }
    const value = state[exports.CODEX_TOPOLOGY_RECEIPTS_KEY];
    const errors = validateCodexTopologyReceipts(value);
    if (!Array.isArray(value)) {
        return errors;
    }
    const receipts = validReceipts(value);
    if (requiredRootPersona !== undefined &&
        !receipts.some((receipt) => receipt.root_persona === requiredRootPersona)) {
        errors.push(`Checkpoint ${exports.CODEX_TOPOLOGY_RECEIPTS_KEY} is missing the forced ` +
            `root persona receipt for ${requiredRootPersona}.`);
    }
    const childReceipts = receipts.filter((receipt) => receipt.root_persona === null);
    if (delegated.size > 0 && childReceipts.length === 0) {
        errors.push(`Checkpoint ${exports.CODEX_TOPOLOGY_RECEIPTS_KEY} is missing a child ` +
            "topology receipt for recorded delegations.");
    }
    for (const receipt of childReceipts) {
        const selectedRoute = state["path_selected"];
        if (receipt.execution_context === "standalone" &&
            (selectedRoute === "small" || selectedRoute === "large") &&
            receipt.route !== selectedRoute) {
            errors.push(`Checkpoint path_selected ${pythonRepr(selectedRoute)} does not match ` +
                `the resolved Codex topology route ${pythonRepr(receipt.route)}.`);
        }
        const allowedNames = modelDeployments(state, receipt.logical_agent);
        allowedNames.add(receipt.logical_agent);
        if ([...delegated].every((agent) => !allowedNames.has(agent))) {
            errors.push("Checkpoint delegation_receipts is missing the exact resolved " +
                `topology agent for ${receipt.logical_agent}: ` +
                `${[...allowedNames].sort().join(", ")}.`);
        }
    }
    return errors;
}
/** Apply always-on present-receipt validation and the optional topology gate. */
function validateCodexTopologyState(state, requireGate = false, options = {}) {
    const errors = [];
    if (exports.CODEX_TOPOLOGY_RECEIPTS_KEY in state) {
        errors.push(...validateCodexTopologyReceipts(state[exports.CODEX_TOPOLOGY_RECEIPTS_KEY]));
    }
    if (requireGate) {
        errors.push(...validateCodexTopologyGate(state, options));
    }
    return errors;
}

File: scripts/dev_tools/_orchestrator_state_codex_topology.py
"""Validate deterministic Codex topology receipts in checkpoints."""

from __future__ import annotations

from typing import Any, TypedDict, cast

from scripts.dev_tools._orchestrator_state_codex_model_routing import (
    CODEX_MODEL_ROUTING_RECEIPTS_KEY,
    validate_codex_model_routing_receipts,
)
from scripts.dev_tools.resolve_codex_topology import (
    FORCED_ROOT_PERSONAS,
    CodexTopologyReceipt,
    resolve_codex_topology,
)

CODEX_TOPOLOGY_RECEIPTS_KEY = "codex_topology_receipts"
_REQUIRED_KEYS = (
    "phase",
    "execution_context",
    "languages",
    "production_file_count",
    "test_file_count",
    "cross_cutting",
    "root_persona",
    "route",
    "topology",
    "logical_agent",
    "routing_reason",
    "max_production_files",
    "max_test_files",
)
_RESOLVED_KEYS = tuple(key for key in _REQUIRED_KEYS if key != "phase")
_REQUIRED_KEY_SET = frozenset(_REQUIRED_KEYS)
_TopologyInputKey = tuple[tuple[str, ...], int, int, str, bool, str | None]


class _TopologyResolverInputs(TypedDict):
    """Type-safe resolver arguments extracted from a checkpoint receipt."""

    languages: list[str]
    production_file_count: int
    test_file_count: int
    execution_context: str
    cross_cutting: bool
    root_persona: str | None


def _receipt_inputs(
    receipt: dict[str, Any], prefix: str
) -> tuple[list[str], _TopologyResolverInputs | None]:
    """Validate resolver input types and return normalized invocation data."""

    errors: list[str] = []
    languages = receipt.get("languages")
    if not isinstance(languages, list) or any(
        not isinstance(language, str) or not language.strip()
        for language in cast("list[object]", languages)
    ):
        errors.append(f"{prefix}.languages must be a list of non-empty strings.")
    for key in ("production_file_count", "test_file_count"):
        value = receipt.get(key)
        if isinstance(value, bool) or not isinstance(value, int):
            errors.append(f"{prefix}.{key} must be an integer.")
    if not isinstance(receipt.get("cross_cutting"), bool):
        errors.append(f"{prefix}.cross_cutting must be a boolean.")
    if not isinstance(receipt.get("execution_context"), str):
        errors.append(f"{prefix}.execution_context must be a string.")
    root_persona = receipt.get("root_persona")
    if root_persona is not None and root_persona not in FORCED_ROOT_PERSONAS:
        errors.append(
            f"{prefix}.root_persona must be null or one of "
            f"{tuple(sorted(FORCED_ROOT_PERSONAS))}."
        )
    if errors:
        return errors, None
    return errors, {
        "languages": cast("list[str]", languages),
        "production_file_count": cast("int", receipt["production_file_count"]),
        "test_file_count": cast("int", receipt["test_file_count"]),
        "execution_context": cast("str", receipt["execution_context"]),
        "cross_cutting": cast("bool", receipt["cross_cutting"]),
        "root_persona": cast("str | None", root_persona),
    }


def validate_codex_topology_receipts(value: object) -> list[str]:
    """Validate every present receipt against the canonical topology resolver."""

    if not isinstance(value, list):
        return [
            f"Checkpoint {CODEX_TOPOLOGY_RECEIPTS_KEY} must be a list when present."
        ]

    errors: list[str] = []
    validated_receipts: dict[_TopologyInputKey, dict[str, object]] = {}
    last_valid_receipt: dict[str, Any] | None = None
    items = cast("list[object]", value)
    for index, item in enumerate(items):
        prefix = f"Checkpoint {CODEX_TOPOLOGY_RECEIPTS_KEY}[{index}]"
        if not isinstance(item, dict):
            errors.append(f"{prefix} must be an object.")
            continue
        receipt = cast("dict[str, Any]", item)
        phase = receipt.get("phase")
        if isinstance(phase, str) and phase.strip() and last_valid_receipt is not None:
            last_valid_receipt["phase"] = phase
            if receipt == last_valid_receipt:
                continue
        if not _REQUIRED_KEY_SET <= receipt.keys():
            missing = [key for key in _REQUIRED_KEYS if key not in receipt]
            errors.append(f"{prefix} missing required keys: {', '.join(missing)}.")
            continue
        if not isinstance(phase, str) or not phase.strip():
            errors.append(f"{prefix}.phase must be a non-empty string.")

        input_errors, inputs = _receipt_inputs(receipt, prefix)
        errors.extend(input_errors)
        if inputs is None:
            continue
        input_key: _TopologyInputKey = (
            tuple(inputs["languages"]),
            inputs["production_file_count"],
            inputs["test_file_count"],
            inputs["execution_context"],
            inputs["cross_cutting"],
            inputs["root_persona"],
        )
        expected_map = validated_receipts.get(input_key)
        if expected_map is None:
            try:
                expected = resolve_codex_topology(**inputs)
            except ValueError as exc:
                errors.append(f"{prefix} has invalid routing inputs: {exc}")
                continue
            expected_map = cast("dict[str, object]", expected)
            validated_receipts[input_key] = expected_map
        matches_expected = True
        for key in _RESOLVED_KEYS:
            if receipt.get(key) != expected_map[key]:
                matches_expected = False
                errors.append(
                    f"{prefix}.{key} must be {expected_map[key]!r}, "
                    f"found {receipt.get(key)!r}."
                )
        if matches_expected and isinstance(phase, str) and phase.strip():
            last_valid_receipt = dict(receipt)
    return errors


def _delegated_agent_names(state: dict[str, Any]) -> set[str]:
    """Collect exact names from legacy or object-form delegation receipts."""

    result: set[str] = set()
    receipts = state.get("delegation_receipts")
    if isinstance(receipts, dict):
        receipts = cast("dict[str, object]", receipts).get("agents")
    if not isinstance(receipts, list):
        return result
    for item in cast("list[object]", receipts):
        if not isinstance(item, dict):
            continue
        agent_name = cast("dict[str, Any]", item).get("agent_name")
        if isinstance(agent_name, str) and agent_name.strip():
            result.add(agent_name)
    return result


def _model_deployments(state: dict[str, Any], logical_agent: str) -> set[str]:
    """Return exact, independently validated deployments for a logical agent."""

    result: set[str] = set()
    receipts = state.get(CODEX_MODEL_ROUTING_RECEIPTS_KEY)
    if not isinstance(receipts, list):
        return result
    for item in cast("list[object]", receipts):
        if not isinstance(item, dict):
            continue
        receipt = cast("dict[str, Any]", item)
        if validate_codex_model_routing_receipts([receipt]):
            continue
        if receipt.get("logical_agent") == logical_agent:
            deployment = receipt.get("deployment_agent")
            if isinstance(deployment, str):
                result.add(deployment)
    return result


def _valid_receipts(value: list[object]) -> list[CodexTopologyReceipt]:
    """Return structurally and deterministically valid receipt objects."""

    result: list[CodexTopologyReceipt] = []
    for item in value:
        if not isinstance(item, dict):
            continue
        receipt = cast("dict[str, Any]", item)
        if validate_codex_topology_receipts([receipt]):
            continue
        result.append(cast("CodexTopologyReceipt", receipt))
    return result


def validate_codex_topology_gate(
    state: dict[str, Any], *, required_root_persona: str | None = None
) -> list[str]:
    """Require topology receipts and exact initial-agent delegations."""

    delegated = _delegated_agent_names(state)
    if not delegated and required_root_persona is None:
        return []

    value = state.get(CODEX_TOPOLOGY_RECEIPTS_KEY)
    errors = validate_codex_topology_receipts(value)
    if not isinstance(value, list):
        return errors
    receipts = _valid_receipts(cast("list[object]", value))

    if required_root_persona is not None and not any(
        receipt["root_persona"] == required_root_persona for receipt in receipts
    ):
        errors.append(
            f"Checkpoint {CODEX_TOPOLOGY_RECEIPTS_KEY} is missing the forced "
            f"root persona receipt for {required_root_persona}."
        )

    non_root_receipts = [
        receipt for receipt in receipts if receipt["root_persona"] is None
    ]
    if delegated and not non_root_receipts:
        errors.append(
            f"Checkpoint {CODEX_TOPOLOGY_RECEIPTS_KEY} is missing a child "
            "topology receipt for recorded delegations."
        )
    for receipt in non_root_receipts:
        logical_agent = receipt["logical_agent"]
        selected_route = state.get("path_selected")
        if (
            receipt["execution_context"] == "standalone"
            and selected_route in {"small", "large"}
            and receipt["route"] != selected_route
        ):
            errors.append(
                f"Checkpoint path_selected {selected_route!r} does not match "
                f"the resolved Codex topology route {receipt['route']!r}."
            )
        allowed_names = {logical_agent} | _model_deployments(state, logical_agent)
        if delegated.isdisjoint(allowed_names):
            errors.append(
                f"Checkpoint delegation_receipts is missing the exact resolved "
                f"topology agent for {logical_agent}: "
                f"{', '.join(sorted(allowed_names))}."
            )
    return errors
